src/python/pants/backend/java/classpath.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass
+from typing import Callable, Iterator
+
+from pants.backend.java.compile.javac import CompiledClassfiles, CompileJavaSourceRequest
+from pants.engine.addresses import Addresses
+from pants.engine.fs import AddPrefix, Digest, MergeDigests, Snapshot
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    CoarsenedTargets,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierLockfileForTargetRequest,
+    CoursierResolvedLockfile,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+)
+
+_USERCP_RELPATH = "__usercp"
+
+
+@dataclass(frozen=True)
+class Classpath:
+    """A transitive classpath which is sufficient to launch the target(s) it was generated for.
+
+    This classpath is guaranteed to contain only JAR files.
+    """
+
+    content: Snapshot
+
+    def classpath_args(self, prefix: str | None = None) -> Iterator[str]:
+        """Construct the argument to be passed to `java -cp`.
+
+        :param prefix: if set, will be prepended to all entries.  This is useful
+            if the process working directory is not the same as the root
+            directory for the process input `Digest`.
+        """
+        return self._classpath(lambda _: True, prefix=prefix)
```

Comment:
I would consider refactoring `_classpath` to be a generator:

```
def _classpath(prefix):
    ...
    ...
    return (maybe_add_prefix(file_path) for file_path in self.content.files)
```

At which point, `classpath_args` becomes a lot simpler just return `_classpath`, and `user_classpath_args` turns into

`return (i for i in self._classpath(prefix) if i.startswith(...)`

I think this will be quite a bit more readable than throwing lambdae around.

Reply:
The lambda needs to run before the prefix is added, so that doesn't quite work.